controllers/awsmachine_controller_unit_test.go

```diff
@@ -17,25 +17,1507 @@ limitations under the License.
+
+					ms.AWSMachine.Spec.AdditionalTags = infrav1.Tags{"kind": "alicorn"}
+					cs.AWSCluster.Spec.AdditionalTags = infrav1.Tags{"colour": "lavender"}
+
+					ec2Svc.EXPECT().UpdateResourceTags(
+						gomock.Any(),
+						map[string]string{
+							"kind": "alicorn",
+						},
+						map[string]string{},
+					).Return(nil).Times(2)
+
+					ec2Svc.EXPECT().UpdateResourceTags(
+						PointsTo("myMachine"),
+						map[string]string{
+							"colour": "lavender",
+							"kind":   "alicorn",
+						},
+						map[string]string{},
+					).Return(nil)
+
+					_, err := reconciler.reconcileNormal(context.Background(), ms, cs, cs, cs)
+					g.Expect(err).To(BeNil())
+				})
+				t.Run("should tag instances volume tags", func(t *testing.T) {
+					g := NewWithT(t)
+					awsMachine := getAWSMachineWithAdditionalTags()
+					setup(t, g, awsMachine)
+					defer teardown(t, g)
+					instanceCreate(t, g)
+					getCoreSecurityGroups(t, g)
+
+					ms.AWSMachine.Spec.AdditionalTags = infrav1.Tags{"rootDeviceID": "id1", "rootDeviceSize": "30"}
+
+					ec2Svc.EXPECT().UpdateResourceTags(
+						gomock.Any(),
+						map[string]string{
+							"rootDeviceID":   "id1",
+							"rootDeviceSize": "30",
+						},
+						map[string]string{},
+					).Return(nil).Times(3)
+
+					_, err := reconciler.reconcileNormal(context.Background(), ms, cs, cs, cs)
+					g.Expect(err).To(BeNil())
+				})
+			})
+
+			t.Run("temporarily stopping then starting the AWSMachine", func(t *testing.T) {
+				var buf *bytes.Buffer
+				getCoreSecurityGroups := func(t *testing.T, g *WithT) {
+					t.Helper()
+
+					buf = new(bytes.Buffer)
+					klog.SetOutput(buf)
+					ec2Svc.EXPECT().GetInstanceSecurityGroups(gomock.Any()).
+						Return(map[string][]string{"eid": {}}, nil).Times(1)
+					secretSvc.EXPECT().UserData(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, nil).Times(1)
+					ec2Svc.EXPECT().GetCoreSecurityGroups(gomock.Any()).Return([]string{}, nil).Times(1)
+					secretSvc.EXPECT().Create(gomock.Any(), gomock.Any()).Return("test", int32(1), nil).Times(1)
+				}
+
+				t.Run("should set instance to stopping and unready", func(t *testing.T) {
+					g := NewWithT(t)
+					awsMachine := getAWSMachine()
+					setup(t, g, awsMachine)
+					defer teardown(t, g)
+					instanceCreate(t, g)
+					getCoreSecurityGroups(t, g)
+
+					instance.State = infrav1.InstanceStateStopping
+					_, _ = reconciler.reconcileNormal(context.Background(), ms, cs, cs, cs)
+					g.Expect(ms.AWSMachine.Status.InstanceState).To(PointTo(Equal(infrav1.InstanceStateStopping)))
+					g.Expect(ms.AWSMachine.Status.Ready).To(Equal(false))
+					g.Expect(buf.String()).To(ContainSubstring(("EC2 instance state changed")))
+					expectConditions(g, ms.AWSMachine, []conditionAssertion{{infrav1.InstanceReadyCondition, corev1.ConditionFalse, clusterv1.ConditionSeverityError, infrav1.InstanceStoppedReason}})
+				})
+
+				t.Run("should then set instance to stopped and unready", func(t *testing.T) {
```

Comment:
All the tests below are stateless:

t.Run("should set instance to stopping and unready"
t.Run("should then set instance to stopped and unready"
t.Run("should then set instance to running and ready once it is restarted"

Each has its own AWSMachine and instance created to test a single scenario, so better be named accordingly.

Reply:
Added a comment that its stateless, not sure how do we rename, wdyt?